File: impl/java/src/main/java/com/minustwo/chanlun/Strokes.java
// Strokes.java — Def-4 + Lemma 2 stroke greedy, faithful port of
// conformance/chanlun-v1/reference_backend/strokes.py:strokes.
//
// Walk the fractal sequence with a single alternating anchor (leftmost-greedy):
//   - same-kind fractal           -> keep the extremal rep (pickRep)
//   - opposite-kind, gap >= dmin  -> EMIT (anchor->f) and re-anchor to f
//   - opposite-kind, gap <  dmin  -> drop as named residue
// dispositions[i] classifies fractal i as one of {first, absorbed, emit, residue}.
package com.minustwo.chanlun;

import java.util.ArrayList;
import java.util.List;

import com.minustwo.chanlun.Types.Fractal;
import com.minustwo.chanlun.Types.OrderedObject;
import com.minustwo.chanlun.Types.Stroke;
import com.minustwo.chanlun.Types.Value;

public final class Strokes {

    private Strokes() {}

    /** Extremal representative among same-kind fractals: top -> higher h; bottom -> lower l. */
    private static Fractal pickRep(Fractal cur, Fractal cand) {
        if (Types.FRACTAL_TOP.equals(cur.kind())) {
            return cand.h() > cur.h() ? cand : cur;
        }
        return cand.l() < cur.l() ? cand : cur;
    }

    /** Result holder for the (strokes, dispositions) tuple. */
    public record Result(List<Stroke> strokes, List<String> dispositions) {}

    /** Run the greedy. Returns (strokes, dispositions). */
    public static Result strokes(List<Fractal> fractals, long dmin) {
        Fractal anchor = null;
        List<Stroke> out = new ArrayList<>();
        List<String> disp = new ArrayList<>(fractals.size());
        for (Fractal f : fractals) {
            if (anchor == null) {
                anchor = f;
                disp.add(Types.DISP_FIRST);
                continue;
            }
            if (f.kind().equals(anchor.kind())) {
                anchor = pickRep(anchor, f);
                disp.add(Types.DISP_ABSORBED);
                continue;
            }
            // opposite kind
            if (f.idx() - anchor.idx() >= dmin) {
                long aPrice = Types.FRACTAL_TOP.equals(anchor.kind()) ? anchor.h() : anchor.l();
                long fPrice = Types.FRACTAL_TOP.equals(f.kind()) ? f.h() : f.l();
                String dir = Types.FRACTAL_TOP.equals(f.kind()) ? "up" : "down";
                out.add(new Stroke(anchor.idx(), f.idx(), dir, aPrice, fPrice));
                anchor = f;
                disp.add(Types.DISP_EMIT);
            } else {
                disp.add(Types.DISP_RESIDUE);
            }
        }
        return new Result(out, disp);
    }

    // ==== Value tree adapters ===============================================

    public static Value strokesToValue(List<Stroke> ss) {
        List<Value> out = new ArrayList<>(ss.size());
        for (Stroke s : ss) {
            OrderedObject o = new OrderedObject();
            o.set("from_idx", Value.ofInt(s.fromIdx()));
            o.set("to_idx", Value.ofInt(s.toIdx()));
            o.set("dir", Value.ofString(s.dir()));
            o.set("from_p", Value.ofInt(s.fromP()));
            o.set("to_p", Value.ofInt(s.toP()));
            out.add(Value.ofObject(o));
        }
        return Value.ofArray(out);
    }

    public static Value dispositionsToValue(List<String> disp) {
        List<Value> out = new ArrayList<>(disp.size());
        for (String d : disp) out.add(Value.ofString(d));
        return Value.ofArray(out);
    }
}

File: impl/csharp/src/Chanlun/Strokes.cs
// Def-4 + Lemma 2 stroke greedy — faithful port of
// conformance/chanlun-v1/reference_backend/strokes.py:strokes.
//
// Walk the fractal sequence with a single alternating anchor (leftmost-greedy):
//   - same-kind fractal      -> keep extremal rep (pick_rep)
//   - opposite-kind & gap>=dmin -> EMIT (anchor->f) and re-anchor to f
//   - opposite-kind but gap<dmin -> drop as named residue
// Dispositions classify EACH fractal into one of
// {first, absorbed, emit, residue} (exhaustion witness).

namespace Chanlun;

public static class StrokeOps
{
    public static (List<Stroke> strokes, List<string> dispositions) Strokes(
        IReadOnlyList<Fractal> fractals, long dmin)
    {
        var outStrokes = new List<Stroke>();
        var disp = new List<string>(fractals.Count);
        bool hasAnchor = false;
        Fractal anchor = default;
        foreach (var f in fractals)
        {
            if (!hasAnchor)
            {
                anchor = f;
                hasAnchor = true;
                disp.Add(Dispositions.First);
                continue;
            }
            if (f.Kind == anchor.Kind)
            {
                anchor = PickRep(anchor, f);
                disp.Add(Dispositions.Absorbed);
                continue;
            }
            // opposite kind
            if (f.Idx - anchor.Idx >= dmin)
            {
                long aPrice = anchor.Kind == Kinds.Top ? anchor.H : anchor.L;
                long fPrice = f.Kind == Kinds.Top ? f.H : f.L;
                string dir = f.Kind == Kinds.Top ? "up" : "down";
                outStrokes.Add(new Stroke(anchor.Idx, f.Idx, dir, aPrice, fPrice));
                anchor = f;
                disp.Add(Dispositions.Emit);
            }
            else
            {
                disp.Add(Dispositions.Residue);
            }
        }
        return (outStrokes, disp);
    }

    private static Fractal PickRep(Fractal cur, Fractal cand)
    {
        if (cur.Kind == Kinds.Top)
            return cand.H > cur.H ? cand : cur;
        // bottom: lower l wins
        return cand.L < cur.L ? cand : cur;
    }

    // ---- Value-tree converters ------------------------------------------------

    public static Value StrokesToValue(IReadOnlyList<Stroke> ss)
    {
        var arr = new List<Value>(ss.Count);
        foreach (var s in ss)
        {
            var o = new OrderedObject();
            // Insertion order doesn't matter (canonical re-sorts) — match
            // Python field order anyway for readability.
            o.Set("from_idx", Value.Of(s.FromIdx));
            o.Set("to_idx", Value.Of(s.ToIdx));
            o.Set("dir", Value.Of(s.Dir));
            o.Set("from_p", Value.Of(s.FromP));
            o.Set("to_p", Value.Of(s.ToP));
            arr.Add(Value.Of(o));
        }
        return Value.Of(arr);
    }

    public static Value DispositionsToValue(IReadOnlyList<string> disp)
    {
        var arr = new List<Value>(disp.Count);
        foreach (var d in disp)
            arr.Add(Value.Of(d));
        return Value.Of(arr);
    }
}
